math/minuit2/src/NumericalDerivator.cxx

```diff
@@ -0,0 +1,336 @@
+ *  Modified version created on: Sep 27, 2017
+ *      Author: E. G. P. Bos
+ *
+ *      NumericalDerivator was essentially a slightly modified copy of code
+ *      written by M. Winkler, F. James, L. Moneta, and A. Zsenei for Minuit2,
+ *      Copyright (c) 2005 LCG ROOT Math team, CERN/PH-SFT. Original version:
+ *      https://github.com/lmoneta/root/blob/lvmini/math/mathcore/src/NumericalDerivator.cxx
+ *
+ *      This class attempts to more closely follow the Minuit2 implementation.
+ *      Modified things (w.r.t. original) are indicated by MODIFIED.
+ */
+
+#include "Minuit2/NumericalDerivator.h"
+#include <cmath>
+#include <algorithm>
+#include <Math/IFunction.h>
+#include <iostream>
+#include <TMath.h>
+#include <cassert>
+#include "Fit/ParameterSettings.h"
+
+#include <Math/Minimizer.h> // needed here because in Fitter is only a forward declaration
+
+namespace ROOT {
+namespace Minuit2 {
+
+NumericalDerivator::NumericalDerivator(bool always_exactly_mimic_minuit2)
+   : fAlwaysExactlyMimicMinuit2(always_exactly_mimic_minuit2)
+{
+}
+
+NumericalDerivator::NumericalDerivator(double step_tolerance, double grad_tolerance, unsigned int ncycles,
+                                       double error_level, bool always_exactly_mimic_minuit2)
+   : fStepTolerance(step_tolerance), fGradTolerance(grad_tolerance), fNCycles(ncycles), fUp(error_level),
+     fAlwaysExactlyMimicMinuit2(always_exactly_mimic_minuit2)
+{
+}
+
+// deep copy constructor
+NumericalDerivator::NumericalDerivator(const NumericalDerivator &other)
+   : fStepTolerance(other.fStepTolerance), fGradTolerance(other.fGradTolerance), fNCycles(other.fNCycles),
+     fUp(other.fUp), fVal(other.fVal), fVx(other.fVx), fVxExternal(other.fVxExternal), fDfmin(other.fDfmin),
+     fVrysml(other.fVrysml), fPrecision(other.fPrecision), fAlwaysExactlyMimicMinuit2(other.fAlwaysExactlyMimicMinuit2),
+     fVxFValCache(other.fVxFValCache)
+{
+}
+
+void NumericalDerivator::SetStepTolerance(double value)
+{
+   fStepTolerance = value;
+}
+
+void NumericalDerivator::SetGradTolerance(double value)
+{
+   fGradTolerance = value;
+}
+
+void NumericalDerivator::SetNCycles(unsigned int value)
+{
+   fNCycles = value;
+}
+
+void NumericalDerivator::SetErrorLevel(double value)
+{
+   fUp = value;
+}
+
+// This function sets internal state based on input parameters. This state
+// setup is used in the actual (partial) derivative calculations.
+void NumericalDerivator::SetupDifferentiate(const ROOT::Math::IBaseFunctionMultiDim *function, const double *cx,
+                                            const std::vector<ROOT::Fit::ParameterSettings> &parameters)
+{
+   assert(function != nullptr && "function is a nullptr");
+
+   if (fVx.size() != function->NDim()) {
+      fVx.resize(function->NDim());
+   }
+   if (fVxExternal.size() != function->NDim()) {
+      fVxExternal.resize(function->NDim());
+   }
+   if (fVxFValCache.size() != function->NDim()) {
+      fVxFValCache.resize(function->NDim());
+   }
```

Comment:
```suggestion
   fVx.resize(function->NDim());
   fVxExternal.resize(function->NDim());
   fVxFValCache.resize(function->NDim());
```
I think it's more readable without this size check, and the `std::vector<T>::resize()` method probably already knows how to handle the trivial case of no resizing.

Reply:
Ah, that's a good point, hadn't thought of that. I guess it'd be implementation dependent, but indeed, makes the code cleaner and it was probably a premature optimization anyway.